### DISCUSSION TITLE
Enum Parameter

### DISCUSSION THREAD

**Discussion by bicarlsen** [pymeasure/pymeasure#663] [Q&A]
Is there a recommended way for encapsulation of an enum parameter?

**Comment by msmttchr** [comment]
I am not sure if there's any recommendations, but I see https://docs.python.org/3/library/enum.html used in some code

**Comment by BenediktBurger** [comment]
Or you can use dictionaries, which are used in other instrument implementations.

**Comment by bicarlsen** [comment]
@bmoneke and @msmttchr Could you point me to a piece of code as an example?

**Comment by LongnoseRob** [comment]
@bicarlsen 
I have used InfFlags for the Error bit and SRQ mask implementation of [this instrument](https://github.com/pymeasure/pymeasure/blob/master/pymeasure/instruments/hp/hp3478A.py#L195-#L214)
another exmaple would be [here](https://github.com/pymeasure/pymeasure/blob/8122c28aa04ba8ec6ca51b6ebf83eb43771aba93/pymeasure/instruments/srs/sr830.py#L34-#L57)

`IntFlags` can be a a good and simple solution when a ctypes bitfield would be overkill.

**Comment by bicarlsen** [comment]
@LongnoseRob, do you then use those flags in an `IntegerParameter`?

**Comment by LongnoseRob** [comment]
> @LongnoseRob, do you then use those flags in an `IntegerParameter`?

yes, example based on `HP3478A.SRQ`:

```
#definition
 class SRQ(IntFlag):
        """Enum element for SRQ mask bit decoding
        """
        Power_on = 128
        Calibration = 32
        Front_panel_button = 16
        Internal_error = 8
        Syntax_error = 4
        Data_ready = 1
```
Depending on the value selected, the instrument would create an SRQ (aka message to GPIB-controller, I have a status, of wihich you wanted me to tell")

simple examples
```
n SRQ condition(s)
1 SRQ.Data_ready
4 SRQ.Syntax_error
5 SRQ.Syntax_error|Data_ready 
8 SRQ.Internal_error
9 SRQ.Internal_error|Data_ready
12 SRQ.Internal_error|Syntax_error
13 SRQ.Internal_error|Syntax_error|Data_ready
24 SRQ.Front_panel_button|Internal_error
25 SRQ.Front_panel_button|Internal_error|Data_ready
28 SRQ.Front_panel_button|Internal_error|Syntax_error
29 SRQ.Front_panel_button|Internal_error|Syntax_error|Data_ready
```

I hope this helps, if not pelase let me know

**Comment by bilderbuchi** [comment]
As this is more of a usage question, I'll convert it into a discussion.
